src/portal/src/app/project/webhook/webhook.component.ts

```diff
@@ -0,0 +1,154 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import { finalize } from "rxjs/operators";
+import { TranslateService } from '@ngx-translate/core';
+import { Component, OnInit, ViewChild } from '@angular/core';
+import { AddWebhookComponent } from "./add-webhook/add-webhook.component";
+import { AddWebhookFormComponent } from "./add-webhook-form/add-webhook-form.component";
+import { ActivatedRoute } from '@angular/router';
+import { Webhook, LastTrigger } from './webhook';
+import { WebhookService } from './webhook.service';
+import { MessageHandlerService } from "../../shared/message-handler/message-handler.service";
+import { Project } from '../project';
+import {
+  ConfirmationTargets,
+  ConfirmationState,
+  ConfirmationButtons
+} from "../../shared/shared.const";
+
+import { ConfirmationMessage } from "../../shared/confirmation-dialog/confirmation-message";
+import { ConfirmationAcknowledgement } from "../../shared/confirmation-dialog/confirmation-state-message";
+import { ConfirmationDialogComponent } from "../../shared/confirmation-dialog/confirmation-dialog.component";
+
+@Component({
+  templateUrl: './webhook.component.html',
+  styleUrls: ['./webhook.component.scss'],
+  // changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class WebhookComponent implements OnInit {
+  @ViewChild(AddWebhookComponent)
+  addWebhookComponent: AddWebhookComponent;
+  @ViewChild(AddWebhookFormComponent)
+  addWebhookFormComponent: AddWebhookFormComponent;
+  @ViewChild("confirmationDialogComponent")
+  confirmationDialogComponent: ConfirmationDialogComponent;
+  webhook: Webhook;
+  endpoint: string = '';
+  lastTriggers: LastTrigger[] = [];
+  lastTriggerCount: number = 0;
+  isEnabled: boolean;
+  loading: boolean = false;
+  showCreate: boolean = false;
+  projectId: number;
+  projectName: string;
+  constructor(
+    private route: ActivatedRoute,
+    private translate: TranslateService,
+    private webhookService: WebhookService,
+    private messageHandlerService: MessageHandlerService) {}
+
+  ngOnInit() {
+    this.projectId = +this.route.snapshot.parent.params['id'];
+    let resolverData = this.route.snapshot.parent.data;
+    if (resolverData) {
+      let project = <Project>(resolverData["projectResolver"]);
+      this.projectName = project.name;
+    }
+    this.getData(this.projectId);
+  }
+
+  getData(projectId: number) {
+    this.getLastTriggers(projectId);
+    this.getWebhook(projectId);
+  }
+
+  getLastTriggers(projectId: number) {
+    this.webhookService
+      .listLastTrigger(projectId)
+      .pipe(finalize(() => (this.loading = false)))
+      .subscribe(
+        response => {
+          this.lastTriggers = response;
+          this.lastTriggerCount = response.length;
+        },
+        error => {
+          this.messageHandlerService.handleError(error);
+        }
+      );
+  }
+
+  getWebhook(projectId: number) {
+    this.loading = true;
```

Comment:
when to set this.loading to false?

Reply:
It's my fault to set `this.loading` in Function `getWebhook`.Because `loading` only used By template of list.I'll set `this.loading` to `true` in Function `getLastTriggers` .